### PR TITLE
Generating Parameterized SQL

### PR DESCRIPTION
**Pull request by sspekinc**
Converting constant expressions to the wrapped ones will enable the generated SQL is to be parameterized as proposed in the following blog post.

https://github.com/graeme-hill/gblog/blob/master/source_content/articles/2014.139_entity-framework-dynamic-queries-and-parameterization.mkd

### CODE REVIEW

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 1]
```diff
@@ -68,12 +68,20 @@ public static Expression GenerateSubtract(Expression left, Expression right)
         public static Expression GenerateEqual(Expression left, Expression right)
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: I think the `WrapConstantExpression` logic should also be applied to the `GenerateStringConcat` method.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 2]
```diff
@@ -195,5 +215,52 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
         {
             return Expression.Call(null, GetStaticMethod(methodName, left, right), new[] { left, right });
         }
+        
+       static void WrapConstantExpression(ref Expression expression)
+        {
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: Do a format-file for this file to align the `{`.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 3]
```diff
@@ -195,5 +215,52 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
         {
             return Expression.Call(null, GetStaticMethod(methodName, left, right), new[] { left, right });
         }
+        
+       static void WrapConstantExpression(ref Expression expression)
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: Do not use oneliner `if`. But always use `{` and `}`.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 4]
```diff
@@ -195,5 +215,52 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
...
+            {
+                NewExpression newExpression = expression as NewExpression;
+
+                if (newExpression.Type == typeof(Guid)) expression = WrappedConstant((Expression.Lambda<Func<Guid>>(newExpression).Compile())());
+
+                if (newExpression.Type == typeof(DateTime)) expression = WrappedConstant((Expression.Lambda<Func<DateTime>>(newExpression).Compile())());
+            }
+        }
+
+        public static MemberExpression WrappedConstant<TValue>(TValue value)
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: This can be private

Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: Add comment here how this is working. (Maybe add the link to that article, or maybe better to also add the article as file to this git.

**Comment by Jo1nes** [comment]
Hi, I have the same requirement and I very glad there is already a fix. Many thanks! Can you tell me if this will be included in the next release?

**Comment by StefH** [comment]
Hello @sspekinc and @Jo1nes,

There are some code review comments and 3 unit-tests fail.
See https://ci.appveyor.com/project/StefH/system-linq-dynamic-core/build/1.0.8.576 for more details.

**Comment by sspekinc** [comment]
Hi Stef,

It seems I need to change the test itself. Shall I do it?

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 5]
```diff
@@ -195,5 +215,70 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
...
+
+            if (expression is ConstantExpression)
+            {
+                ConstantExpression constantExpression = expression as ConstantExpression;
+
+                if (constantExpression.Type == typeof(string)) 
+                {
+                    expression = WrappedConstant((string)constantExpression.Value);
+                }
+                else if (constantExpression.Type == typeof(Int64))
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: You can use `long` here.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 6]
```diff
@@ -195,5 +215,70 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
...
+
+                if (constantExpression.Type == typeof(string)) 
+                {
+                    expression = WrappedConstant((string)constantExpression.Value);
+                }
+                else if (constantExpression.Type == typeof(Int64)) 
+                {
+                    expression = WrappedConstant((Int64)constantExpression.Value);
+                } 
+                else if (constantExpression.Type == typeof(Int32))
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]: You can use `int` here.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs [hunk 7]
```diff
@@ -195,5 +215,70 @@ static Expression GenerateStaticMethodCall(string methodName, Expression left, E
...
+                }
+                else if (constantExpression.Type == typeof(Int64)) 
+                {
+                    expression = WrappedConstant((Int64)constantExpression.Value);
+                } 
+                else if (constantExpression.Type == typeof(Int32)) 
+                {
+                    expression = WrappedConstant((Int32)constantExpression.Value);
+                }
+                else if (constantExpression.Type == typeof(Int16))
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]:
```suggestion
                else if (constantExpression.Type == typeof(short)) 
```

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs (hunk 6 above)
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]:
```suggestion
                else if (constantExpression.Type == typeof(int)) 
```

File: src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs (hunk 5 above)
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionHelper.cs]:
```suggestion
                else if (constantExpression.Type == typeof(long)) 
```

**Comment by StefH** [comment]
I've refactored some code.
However when I run the `ConsoleApp_net452_EF6` example, the SQL profiler shows the same trace for all 3 queries:

```
exec sp_executesql N'SELECT TOP (1) 
    [Extent1].[Id] AS [Id], 
    [Extent1].[EmployeeNumber] AS [EmployeeNumber], 
    [Extent1].[FirstName] AS [FirstName], 
    [Extent1].[LastName] AS [LastName], 
    [Extent1].[Email] AS [Email], 
    [Extent1].[HireDate] AS [HireDate], 
    [Extent1].[CompanyId] AS [CompanyId], 
    [Extent1].[CountryId] AS [CountryId], 
    [Extent1].[FunctionId] AS [FunctionId], 
    [Extent1].[SubFunctionId] AS [SubFunctionId], 
    [Extent1].[Assigned] AS [Assigned]
    FROM [dbo].[KendoGrid_Employee] AS [Extent1]
    WHERE [Extent1].[EmployeeNumber] > @p__linq__0',N'@p__linq__0 int',@p__linq__0=1000
```

```
exec sp_executesql N'SELECT TOP (1) 
    [Extent1].[Id] AS [Id], 
    [Extent1].[EmployeeNumber] AS [EmployeeNumber], 
    [Extent1].[FirstName] AS [FirstName], 
    [Extent1].[LastName] AS [LastName], 
    [Extent1].[Email] AS [Email], 
    [Extent1].[HireDate] AS [HireDate], 
    [Extent1].[CompanyId] AS [CompanyId], 
    [Extent1].[CountryId] AS [CountryId], 
    [Extent1].[FunctionId] AS [FunctionId], 
    [Extent1].[SubFunctionId] AS [SubFunctionId], 
    [Extent1].[Assigned] AS [Assigned]
    FROM [dbo].[KendoGrid_Employee] AS [Extent1]
    WHERE [Extent1].[EmployeeNumber] > @p__linq__0',N'@p__linq__0 int',@p__linq__0=1001
```

```
exec sp_executesql N'SELECT TOP (1) 
    [Extent1].[Id] AS [Id], 
    [Extent1].[EmployeeNumber] AS [EmployeeNumber], 
    [Extent1].[FirstName] AS [FirstName], 
    [Extent1].[LastName] AS [LastName], 
    [Extent1].[Email] AS [Email], 
    [Extent1].[HireDate] AS [HireDate], 
    [Extent1].[CompanyId] AS [CompanyId], 
    [Extent1].[CountryId] AS [CountryId], 
    [Extent1].[FunctionId] AS [FunctionId], 
    [Extent1].[SubFunctionId] AS [SubFunctionId], 
    [Extent1].[Assigned] AS [Assigned]
    FROM [dbo].[KendoGrid_Employee] AS [Extent1]
    WHERE [Extent1].[EmployeeNumber] > @p__linq__0',N'@p__linq__0 int',@p__linq__0=1002
```

So it seems the code is not working ?

**Comment by sspekinc** [comment]
Seems good. What's next?

**Comment by StefH** [comment]
@sspekinc The generated SQL does **not**  look good.

I expected something like:
``` sql
WHERE [Extent1].[EmployeeNumber] > 1002
```
instead of
``` sql
WHERE [Extent1].[EmployeeNumber] > @p__linq__0',N'@p__linq__0 int',@p__linq__0=1002
```

**Comment by sspekinc** [comment]
I think it should be the default behavior but it would be good to be able to turn it off through ParsingConfig.

**Comment by StefH** [comment]
Added config setting.

Closing this PR ; somehow I need to use my own PR for merging...

**Pull request closed without merging**